### PR TITLE
Do not promote parameters from the parent constructor

### PR DESCRIPTION
**Pull request by mamazu**
This stops the promotion of parameters in the constructor that have the same name as the properties in the parent constructor. This does not work when the names differ.

Refs #2119

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,7 @@ Improvements:
 
   - Improve Diagnostics: Run linters in parallel #2327
   - Index documents on save #2326
+  - Do not promote parameters that are used in parent constructor #2119
```
Comment (dantleech) [inline CHANGELOG.md:10]: this is the first feature in a new release now, you can add a new `## master` section :)

**Review by dantleech — APPROVED** [review]

**Comment by mamazu** [comment]
I moved the changelog entry to the master, is the implementation of "parameters with the same name" good enough? Or do you have a better idea?

File: lib/CodeTransform/Adapter/WorseReflection/Transformer/CompleteConstructor.php [hunk 2]
```diff
@@ -130,14 +130,41 @@ private function transformPromote(SourceCode $source): TextEdits
...
         return TextEdits::fromTextEdits($edits);
     }
 
+    /**
+    * Get the names of the parent constructor to know which parameters should not be promoted.
+    * @return array<string>
+    */
+    private function getParentClassParamaterNames(ReflectionClass $class): array
+    {
+        $ancestor = $class->ancestors()->firstOrNull();
```
Comment (dantleech) [inline lib/CodeTransform/Adapter/WorseReflection/Transformer/CompleteConstructor.php:151]: `parent()`? ancestors includes interfaces etc.

Comment (mamazu) [reply lib/CodeTransform/Adapter/WorseReflection/Transformer/CompleteConstructor.php]: Well we probably need a loop in here to see if the classes even higher up have constructors as well like so:
```php
class B {
    public function __construct(private string $a) {}
}
class A extends B {}

class Foo extends A {
    public function __construct(string $a) {parent::__construct($a);}
}
```

Then it wouldn't pick up on the constructor right?

Comment (dantleech) [reply lib/CodeTransform/Adapter/WorseReflection/Transformer/CompleteConstructor.php]: If you call `parent()->methods()->get('__construct')` it should return the first constructor in the heirarchy starting from the parent

Comment (mamazu) [reply lib/CodeTransform/Adapter/WorseReflection/Transformer/CompleteConstructor.php]: Added a test for it and it seems to pass.

**Comment by dantleech** [comment]
Ta!

**Comment by mamazu** [comment]
You also want to close the issue for that? #2119

**Comment by dantleech** [comment]
Yep thanks

**Pull request merged**